### PR TITLE
Working on text colspan and rowspan.

### LINKED ISSUE #2915: Support row/column spanning tables in text build
When trying to build text docs with row/column spanning tables (#28), I get the following `NotImplementedError`:

`Column or row spanning cells are not implemented.`

With the error, the build is aborted.

### PR DESCRIPTION
**Pull request by JulienPalard**
### Feature or Bugfix
- Both :]

### Purpose
Implement table colspan and rowspan on text writer.

### Relates
fixes #2915

### CODE REVIEW

**Comment by JulienPalard** [comment]
Maybe it can gain some comments, wait for it...

**Comment by JulienPalard** [comment]
Added a commit with comments to explain how this blob works.

File: sphinx/writers/text.py [hunk 1]
```diff
@@ -29,6 +31,235 @@
...
+    Cell spanning on multiple rows or multiple columns (having a
+    colspan or rowspan greater than one) are automatically referenced
+    by all the table cells they covers. This is a usefull
+    representation as we can simply check `if self[x, y] is self[x,
+    y+1]` to recognize a rowspan.
+
+    Colwidth is not automatically computed, it has to be given, either
+    at construction time, either during the table construction.
+
+    Example usage:
```
Comment (matrixise) [inline sphinx/writers/text.py]: not with `.. doctest::` ou `::` ?

File: sphinx/writers/text.py [hunk 2]
```diff
@@ -29,6 +31,235 @@
...
+    | foo    | bar    |
+    |========|========|
+    | FOO    | BAR    |
+    +--------+--------+
+
+    """
+    def __init__(self, colwidth=None):
+        self.lines = []
+        self.separator = 0
+        if colwidth is None:
```
Comment (matrixise) [inline sphinx/writers/text.py]:
```python
self.colwidth = [] if colwidth is None else coldwidth
```

File: sphinx/writers/text.py [hunk 3]
```diff
@@ -29,6 +31,235 @@
...
+        while len(self.lines) < line:
+            self.lines.append([])
+
+    def _ensure_has_column(self, col):
+        for line in self.lines:
+            while len(line) < col:
+                line.append(None)
+
+    def __repr__(self):
+        out = []
```
Comment (matrixise) [inline sphinx/writers/text.py]: use a generator expression, it's more efficient. 
```
return '\n'.join(repr(line) for line in self.lines)
```

File: sphinx/writers/text.py (hunk 1 above)
Comment (JulienPalard) [reply sphinx/writers/text.py]: Oh, I'm surprised they use rst in their docstring. Why not.

File: sphinx/writers/text.py (hunk 3 above)
Comment (JulienPalard) [reply sphinx/writers/text.py]: 👍

File: tests/roots/test-build-text/table_colspan.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+        +-----+-----+
+        | XXX | XXX |
+        +-----+-----+
+        |     | XXX |
+        +-----+     |
+        | XXX |     |
+        +-----+-----+
```
Comment (tk0miya) [inline tests/roots/test-build-text/table_colspan.txt]: Why are these tables indented?

**Review by tk0miya — APPROVED** [review]
Great work!
LGTM with nits.
And please add your name to AUTHORS file :-)

Thanks,

File: tests/roots/test-build-text/table_colspan.txt (hunk 4 above)
Comment (JulienPalard) [reply tests/roots/test-build-text/table_colspan.txt]: table.txt was indented so I blindly indented mines too. Removed the indentation and it still work \o/

**Review by tk0miya — APPROVED** [review]
Thank you for update. LGTM!

**Comment by tk0miya** [comment]
Merged. Thank you for your great work!

**Pull request merged**